### PR TITLE
ForCtorParam MapFrom any expression

### LINKED ISSUE #3159: MapFrom doesn't resolve constructor parameters that have no matching source property
Re-opening issue #2745 with additional details.

### Source/destination types

```csharp
public class Source
{
    public string Result { get; }

    public Source(string result)
    {
        Result = result;
    }
}

public class Dest
{
    public string Result{ get; }
    public dynamic Details { get; }

    public Dest(string result, DestInner1 inner1)
    {
        Result = result;
        Details = inner1;
    }
    public Dest(string result, DestInner2 inner2)
    {
        Result = result;
        Details = inner2;
    }

    public class DestInner1
    {
        public int Value { get; }

        public DestInner1(int value)
        {
            Value = value;
        }
    }

    public class DestInner2
    {
        public int Value { get; }

        public DestInner2(int value)
        {
            Value = value;
        }
    }
}
```

### Mapping configuration

```csharp
protected override MapperConfiguration Configuration { get; } = new MapperConfiguration(config =>
{
    config.CreateMap<Source, Dest>()
        .ForCtorParam("inner2", cfg => cfg.MapFrom(_ => new Dest.DestInner2(100)));
});
```

### Expected behavior

Destination.Details should contain an instance of DestInner2.

### Actual behavior

AutoMapper.AutoMapperConfigurationException: 'Dest does not have a constructor with a parameter named 'inner2'.

### Steps to reproduce

```csharp
public void Should_redirect_value()
{
    var dest = Mapper.Map<Source, Dest>(new Source("Success"));

    dest.ShouldNotBeNull();
}
```

I have been looking into this issue, and following the comments in #2745, I added a dummy "inner2" parameter to the `Source` object, which causes `MapDestinationPropertyToSource` in `TypeMapFactory` to pass.

However, because `Dest` has two constructors, one accepting an instance of `inner1` and the other `inner2`, I would need to have both dummy parameters in `Source`.

Unfortunately, due to the way that constructor mapping is achieved, the first constructor always passes, so it always hits the "cannot find a constructor with parameter 'inner2'" error.

I note that the previous person to report this was asked to submit a PR last August and hasn't done so, so I have looked into it myself.  I have a working prototype, but I don't know the codebase anywhere near well enough to know if I'm missing things.

All existing tests still pass, as does the one I have added to Constructor.cs.

Any assistance would be appreciated.

### PR DESCRIPTION
**Pull request by spydacarnage**
Fixes #3159.

### CODE REVIEW

File: src/AutoMapper/ConstructorMap.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System.Linq.Expressions;
 using System.Reflection;
 using AutoMapper.Execution;
+using AutoMapper.Configuration;
```
Comment (lbargaoanu) [inline src/AutoMapper/ConstructorMap.cs]: Please revert changes to this file.

File: src/AutoMapper/ProfileMap.cs [hunk 2]
```diff
@@ -32,8 +32,8 @@ public ProfileMap(IProfileConfiguration profile, IConfiguration configuration)
             EnableNullPropagationForQueryMapping = profile.EnableNullPropagationForQueryMapping ?? configuration?.EnableNullPropagationForQueryMapping ?? false;
```
Comment (lbargaoanu) [inline src/AutoMapper/ProfileMap.cs]: Please revert changes to this file.

File: src/AutoMapper/TypeMapFactory.cs [hunk 3]
```diff
@@ -24,52 +24,52 @@ public TypeMap CreateTypeMap(Type sourceType, Type destinationType, ProfileMap o
...
-            {
-                foreach (var destCtor in destTypeInfo.Constructors.OrderByDescending(ci => ci.GetParameters().Length))
-                {
-                    if (MapDestinationCtorToSource(typeMap, destCtor, sourceTypeInfo, options))
-                    {
-                        break;
-                    }
-                }
-            }
+            //if (!destinationType.IsAbstract())
```
Comment (lbargaoanu) [inline src/AutoMapper/TypeMapFactory.cs]: Remove this.

File: src/AutoMapper/TypeMapFactory.cs [hunk 4]
```diff
@@ -24,52 +24,52 @@ public TypeMap CreateTypeMap(Type sourceType, Type destinationType, ProfileMap o
...
             {
                 return false;
             }
             return options.MemberConfigurations.Any(_ => _.MapDestinationPropertyToSource(options, sourceTypeInfo, destType, destMemberType, destMemberInfo, members));
         }
 
-        private bool MapDestinationCtorToSource(TypeMap typeMap, ConstructorInfo destCtor, TypeDetails sourceTypeInfo, ProfileMap options)
-        {
-            var ctorParameters = destCtor.GetParameters();
+        //private bool MapDestinationCtorToSource(TypeMap typeMap, ConstructorInfo destCtor, TypeDetails sourceTypeInfo, ProfileMap options)
```
Comment (lbargaoanu) [inline src/AutoMapper/TypeMapFactory.cs]: Remove this.

File: src/AutoMapper/ConstructorMap.cs (hunk 1 above)
Comment (spydacarnage) [reply src/AutoMapper/ConstructorMap.cs]: Updates made

File: src/AutoMapper/ProfileMap.cs (hunk 2 above)
Comment (spydacarnage) [reply src/AutoMapper/ProfileMap.cs]: Updates made

File: src/AutoMapper/TypeMapFactory.cs (hunk 3 above)
Comment (spydacarnage) [reply src/AutoMapper/TypeMapFactory.cs]: Updates made

File: src/AutoMapper/TypeMapFactory.cs (hunk 4 above)
Comment (spydacarnage) [reply src/AutoMapper/TypeMapFactory.cs]: Updates made

File: src/UnitTests/General.cs [hunk 5]
```diff
@@ -3,11 +3,41 @@
 using Shouldly;
```
Comment (lbargaoanu) [inline src/UnitTests/General.cs]: Revert changes to this file.

Comment (spydacarnage) [reply src/UnitTests/General.cs]: Apologies - that was a hangover from testing another reported issue simultaneously.  Removed now.

File: src/AutoMapper/Configuration/MappingExpressionBase.cs [hunk 6]
```diff
@@ -100,6 +113,41 @@ public void Configure(TypeMap typeMap)
...
+                }
+                ctorMap.AddParameter(parameter, resolvers.ToArray(), canResolve);
+            }
+
+            typeMap.ConstructorMap = ctorMap;
+
+            return ctorMap.CanResolve;
+        }
+
+        private bool MapDestinationPropertyToSource(ProfileMap options, TypeDetails sourceTypeInfo, Type destType, Type destMemberType, string destMemberInfo, LinkedList<MemberInfo> members)
```
Comment (lbargaoanu) [inline src/AutoMapper/Configuration/MappingExpressionBase.cs]: That seems just copy pasted from TypeMapFactory.

Comment (spydacarnage) [reply src/AutoMapper/Configuration/MappingExpressionBase.cs]: It is, as it is needed in both locations.  I wasn't sure what your policy on code sharing is.

I could remove it from both locations and add it as an instance method of `ProfileMap`, as that is shared between both locations?  (I would also move `MapDestinationCtorToSource` to `ProfileMap`, too, as they are obviously linked).

Thoughts?

Comment (lbargaoanu) [reply src/AutoMapper/Configuration/MappingExpressionBase.cs]: OK, move them both.

File: src/AutoMapper/Configuration/CtorParamConfigurationExpression.cs [hunk 7]
```diff
@@ -43,5 +43,10 @@ public void Configure(TypeMap typeMap)
                 action(parameter);
             }
         }
+
+        public bool CheckCtorParamName(string paramName)
```
Comment (lbargaoanu) [inline src/AutoMapper/Configuration/CtorParamConfigurationExpression.cs]: This just hides the meaning. Just expose the parameter name.

File: src/AutoMapper/Configuration/MappingExpressionBase.cs (hunk 6 above)
Comment (spydacarnage) [reply src/AutoMapper/Configuration/MappingExpressionBase.cs]: Moved

File: src/AutoMapper/Configuration/CtorParamConfigurationExpression.cs (hunk 7 above)
Comment (spydacarnage) [reply src/AutoMapper/Configuration/CtorParamConfigurationExpression.cs]: Update made

File: src/UnitTests/Constructors.cs [hunk 8]
```diff
@@ -1609,4 +1609,133 @@ public void Should_redirect_value()
...
+            config.CreateMap<Source, Dest>()
+                .ForCtorParam("inner1", cfg => cfg.MapFrom(_ => new Dest.DestInner1(100)));
+        });
+
+        [Fact]
+        public void Should_redirect_value()
+        {
+            var dest = Mapper.Map<Dest>(new Source("Success"));
+
+            dest.ShouldNotBeNull();
```
Comment (lbargaoanu) [inline src/UnitTests/Constructors.cs]: Assert here what the property values should be. The same for the other test.

Comment (spydacarnage) [reply src/UnitTests/Constructors.cs]: Done

**Pull request merged**